### PR TITLE
Add support for OpenTracing

### PR DESCRIPTION
**Pull request by maskit**
This adds very basic support for [OpenTracing](https://opentracing.io).

- Build option to use OpenTracing
- Settings to enable tracing
- Code to initialize libraries
- Generic interface for tracing
- Code to trace transactions

### CODE REVIEW

**Comment by shukitchan** [comment]
i have been using opentracing in a different context. i.e. in a plugin to allow http requests passing through ATS to participate in end-to-end, distributed tracing with jaeger in microservice environment.
And I have since moved on to investigate using opentelemetry instead. 

If we are to add tracing support inside ATS, should we consider using opentelemetry instead?
https://opensource.googleblog.com/2019/05/opentelemetry-merger-of-opencensus-and.html

**Comment by maskit** [comment]
> i have been using opentracing in a different context. i.e. in a plugin to allow http requests passing through ATS to participate in end-to-end, distributed tracing with jaeger in microservice environment.

I think that is how OpenTracing is used by nginx and envoy, and that would be also nice to support. The library detection part should be useful for that case too if we have a plugin like that in our source tree.

> If we are to add tracing support inside ATS, should we consider using opentelemetry instead?

Adding support for OpenTelemetry is fine, but I'm not sure if we should only support OpenTelemetry. I defined a generic interface so that we can support different tracing systems without adding many `#ifdef` here and there. OpenTelemetry library for C++ made version 1.0 just a few months ago.

**Comment by shukitchan** [comment]
I just have the concern that since the announcement of opentracing and opencensus merging to form opentelemetry in 2019, there are not a lot of meaningful updates on the opentracing github repos. (e,.g. https://github.com/opentracing/opentracing-cpp) . So I wonder if the library is still being maintained or in life support mode. And therefore whether we should still be using it or not.

File: include/tscore/Tracing.h [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+#include "tscore/ink_config.h"
+
+#if TS_USE_OPENTRACING
+#include "tscore/Tracing_opentracing.h"
+#else
+using TRACER = int;
+inline void
+tracing_tag(TRACER &out, std::string_view name, std::string_view value)
+{
+  ink_assert(!"No tracng library is available");
```
Comment (SolidWallOfCode) [inline include/tscore/Tracing.h:39]: "tracng"?

**Comment by bryancall** [comment]
@shukitchan Can you take a look at this?  Might want to review after @maskit gives a talk on this at the Fall Summit next week.

**Comment by maskit** [comment]
> there are not a lot of meaningful updates on the opentracing github repos.

Right, but I wouldn't expect a lot of changes on opentracing-cpp repo, because it just provides a bridge for implementations. At least an actual cpp implementation ([jaeger](https://github.com/jaegertracing/jaeger-client-cpp/releases)) is still maintained, although their documentation recommends using OpenTelemetry for new applications.

**Comment by maskit** [comment]
The slides I used for the Summit last week is on the wiki.
https://cwiki.apache.org/confluence/display/TS/Presentations+-+2021

File: include/tscore/Tracing_opentracing.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+using TRACER = opentracing::Span;
+
+extern opentracing::expected<opentracing::DynamicTracingLibraryHandle, std::error_code> ot_lib;
+extern std::string ot_config;
+extern std::mutex tracers_mutex;
+extern std::vector<std::shared_ptr<opentracing::Tracer>> tracers;
+extern ink_thread_key thread_specific_tracer_key;
+extern std::string ot_config;
+extern opentracing::expected<opentracing::DynamicTracingLibraryHandle, std::error_code> ot_lib;
```
Comment (shukitchan) [inline include/tscore/Tracing_opentracing.h:42]: are these duplicate?

**Review by shukitchan — CHANGES_REQUESTED** [review]
Can we add some documentation, especially on integration with opentracing?

I still think we should use opentelemetry but I guess I can add that implementation relatively easily once this PR is merged.

**Comment by sudheerv** [comment]
+1 on OpenTelemetry (OTEL). OpenTracing and OpenCensus have merged into OTEL and are on path to deprecation, so we may not gain much benefit in adding support for OpenTracing at this point. OTEL is also backward compatible with OpenTracing (and OpenCensus) as well.

Coincidentally, we are adding support for Distributed Tracing internally and are adopting OTEL.

File: include/tscore/Tracing.h [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+inline void
+Tracing::delete_tracer(TRACER *tracer)
+{
+  tracing_delete(tracer);
+}
+
+extern std::unique_ptr<Tracing> tracing;
+
+#define TRACE_TAG(out, category, message)   \
```
Comment (sudheerv) [inline include/tscore/Tracing.h:107]: I think we should have TS API to allow plugins to leverage/add tags too. Using Tracing just within ATS core may not have much value as the Tracing is primarily meant to solve the distributed system use case (trace request as it travels through micro services).

**Comment by maskit** [comment]
I do think it's nice to add support for OpenTelemetry and the distributed system use case in the future, but those are out of my scope. I probably don't work on those until I need those, although I'll be happy to review PRs for those. It should not be too difficult to add those supports on top of this change.

I'm going to add documentation unless you think we should not support OpenTracing, but If you think OpenTelemetry will be one and only distributed tracing system and we won't need a complicated abstraction layer, I'd close this PR.

**Comment by shukitchan** [comment]
I don't think there is any other well-maintained alternative to opentelemetry. Perhaps we should close this PR

**Comment by sudheerv** [comment]
> I do think it's nice to add support for OpenTelemetry and the distributed system use case in the future, but those are out of my scope. I probably don't work on those until I need those, although I'll be happy to review PRs for those. It should not be too difficult to add those supports on top of this change.
> 
> I'm going to add documentation unless you think we should not support OpenTracing, but 

>>> If you think OpenTelemetry will be one and only distributed tracing system and we won't need a complicated abstraction layer, I'd close this PR.

Yeah, this definitely seems to be the current direction for distributed tracing. OpenTracing and OpenCensus are both deprecated and are on path to EOL. Both the communities have joined hands and are focusing on OpenTelemetry which ensures backward compatibility with both and is the one where there is quite a bit of interest and adoption across the industry. Even the most popular Jaeger deprecated their own clients and are recommending to switch to OTEL. As such, we may not want to support OpenTracing at this stage. I’m not strongly opposed to it but I’m not too sure if there’s a lot of value in creating an abstraction layer given that OTEL is the only potential standard for the near future at least.

**Comment by maskit** [comment]
Closing this PR because there's no demand for support for a deprecated library and an abstraction layer to support multiple tracing systems. The consensus here was that we should only support OpenTelemetry without abstraction layers.

**Pull request closed without merging**